### ISSUE TITLE
Build fails with 32bit target

### ISSUE THREAD

**Issue by rnd-ash** [napi-rs/napi-rs#284]
Trying to build for MSVC x86 due to loading 32bit Dlls on WIndows...

This fails with the NAPI library being the cause. Attached is a log of the build process:

Seems to compain a lot about expecting u32 but finding u64

[build_log.txt](https://github.com/napi-rs/napi-rs/files/5487606/build_log.txt)

**Comment by rnd-ash** [comment]
Different errors now!
```
> cargo +stable-i686-pc-windows-msvc build && napi build

warning: crate `J2534Common` should have a snake case name
  |
  = note: `#[warn(non_snake_case)]` on by default
  = help: convert the identifier to snake case: `j2534_common`

warning: method `fromByte` should have a snake case name
  --> Z:\MacchinaM2-J2534-Rust\J2534Common\src\lib.rs:28:12
   |
28 |     pub fn fromByte(protocol_id: u32) -> Option<Protocol> {
   |            ^^^^^^^^ help: convert the identifier to snake case: `from_byte`

warning: 2 warnings emitted

   Compiling h2 v0.2.7
   Compiling napi-sys v0.4.7 (https://github.com/napi-rs/napi-rs#f56852bb)
error: failed to run custom build command for `napi-sys v0.4.7 (https://github.com/napi-rs/napi-rs#f56852bb)`

Caused by:
  process didn't exit successfully: `Z:\app\target\debug\build\napi-sys-9006196585aafa0a\build-script-build` (exit code: 101)
  --- stdout
  cargo:rerun-if-env-changed=NODE_INCLUDE_PATH
  cargo:rerun-if-changed=src\bindings.h
  cargo:rerun-if-changed=src\lib.rs
  cargo:rerun-if-changed=src\sys.rs

  --- stderr
  Z:\app\target\debug\build\napi-sys-9f2ceac139b31565\out\node-v12.19.0\include\node\node_api.h:123:41: error: unknown type name 'napi_async_context'
  Z:\app\target\debug\build\napi-sys-9f2ceac139b31565\out\node-v12.19.0\include\node\node_api.h:126:44: error: unknown type name 'napi_async_context'
  Z:\app\target\debug\build\napi-sys-9f2ceac139b31565\out\node-v12.19.0\include\node\node_api.h:129:44: error: unknown type name 'napi_async_context'
  Z:\app\target\debug\build\napi-sys-9f2ceac139b31565\out\node-v12.19.0\include\node\node_api.h:165:36: error: unknown type name 'napi_async_execute_callback'
  Z:\app\target\debug\build\napi-sys-9f2ceac139b31565\out\node-v12.19.0\include\node\node_api.h:166:36: error: unknown type name 'napi_async_complete_callback'
  Z:\app\target\debug\build\napi-sys-9f2ceac139b31565\out\node-v12.19.0\include\node\node_api.h:168:36: error: unknown type name 'napi_async_work'
  Z:\app\target\debug\build\napi-sys-9f2ceac139b31565\out\node-v12.19.0\include\node\node_api.h:170:48: error: unknown type name 'napi_async_work'
  Z:\app\target\debug\build\napi-sys-9f2ceac139b31565\out\node-v12.19.0\include\node\node_api.h:172:47: error: unknown type name 'napi_async_work'
  Z:\app\target\debug\build\napi-sys-9f2ceac139b31565\out\node-v12.19.0\include\node\node_api.h:174:48: error: unknown type name 'napi_async_work'
  Z:\app\target\debug\build\napi-sys-9f2ceac139b31565\out\node-v12.19.0\include\node\node_api.h:179:41: error: unknown type name 'napi_node_version'
  Z:\app\target\debug\build\napi-sys-9f2ceac139b31565\out\node-v12.19.0\include\node\node_api.h:203:50: error: unknown type name 'napi_async_context'
  Z:\app\target\debug\build\napi-sys-9f2ceac139b31565\out\node-v12.19.0\include\node\node_api.h:204:50: error: unknown type name 'napi_callback_scope'
  Z:\app\target\debug\build\napi-sys-9f2ceac139b31565\out\node-v12.19.0\include\node\node_api.h:207:51: error: unknown type name 'napi_callback_scope'
  Z:\app\target\debug\build\napi-sys-9f2ceac139b31565\out\node-v12.19.0\include\node\node_api.h:225:33: error: unknown type name 'napi_threadsafe_function_call_js'
  Z:\app\target\debug\build\napi-sys-9f2ceac139b31565\out\node-v12.19.0\include\node\node_api.h:226:33: error: unknown type name 'napi_threadsafe_function'
  Z:\app\target\debug\build\napi-sys-9f2ceac139b31565\out\node-v12.19.0\include\node\node_api.h:229:38: error: unknown type name 'napi_threadsafe_function'
  Z:\app\target\debug\build\napi-sys-9f2ceac139b31565\out\node-v12.19.0\include\node\node_api.h:233:31: error: unknown type name 'napi_threadsafe_function'
  Z:\app\target\debug\build\napi-sys-9f2ceac139b31565\out\node-v12.19.0\include\node\node_api.h:235:31: error: unknown type name 'napi_threadsafe_function_call_mode'
  Z:\app\target\debug\build\napi-sys-9f2ceac139b31565\out\node-v12.19.0\include\node\node_api.h:238:34: error: unknown type name 'napi_threadsafe_function'
  fatal error: too many errors emitted, stopping now [-ferror-limit=]
  Z:\app\target\debug\build\napi-sys-9f2ceac139b31565\out\node-v12.19.0\include\node\node_api.h:123:41: error: unknown type name 'napi_async_context', err: true
  Z:\app\target\debug\build\napi-sys-9f2ceac139b31565\out\node-v12.19.0\include\node\node_api.h:126:44: error: unknown type name 'napi_async_context', err: true
  Z:\app\target\debug\build\napi-sys-9f2ceac139b31565\out\node-v12.19.0\include\node\node_api.h:129:44: error: unknown type name 'napi_async_context', err: true
  Z:\app\target\debug\build\napi-sys-9f2ceac139b31565\out\node-v12.19.0\include\node\node_api.h:165:36: error: unknown type name 'napi_async_execute_callback', err: true
  Z:\app\target\debug\build\napi-sys-9f2ceac139b31565\out\node-v12.19.0\include\node\node_api.h:166:36: error: unknown type name 'napi_async_complete_callback', err: true
  Z:\app\target\debug\build\napi-sys-9f2ceac139b31565\out\node-v12.19.0\include\node\node_api.h:168:36: error: unknown type name 'napi_async_work', err: true
  Z:\app\target\debug\build\napi-sys-9f2ceac139b31565\out\node-v12.19.0\include\node\node_api.h:170:48: error: unknown type name 'napi_async_work', err: true
  Z:\app\target\debug\build\napi-sys-9f2ceac139b31565\out\node-v12.19.0\include\node\node_api.h:172:47: error: unknown type name 'napi_async_work', err: true
  Z:\app\target\debug\build\napi-sys-9f2ceac139b31565\out\node-v12.19.0\include\node\node_api.h:174:48: error: unknown type name 'napi_async_work', err: true
  Z:\app\target\debug\build\napi-sys-9f2ceac139b31565\out\node-v12.19.0\include\node\node_api.h:179:41: error: unknown type name 'napi_node_version', err: true
  Z:\app\target\debug\build\napi-sys-9f2ceac139b31565\out\node-v12.19.0\include\node\node_api.h:203:50: error: unknown type name 'napi_async_context', err: true
  Z:\app\target\debug\build\napi-sys-9f2ceac139b31565\out\node-v12.19.0\include\node\node_api.h:204:50: error: unknown type name 'napi_callback_scope', err: true
  Z:\app\target\debug\build\napi-sys-9f2ceac139b31565\out\node-v12.19.0\include\node\node_api.h:207:51: error: unknown type name 'napi_callback_scope', err: true
  Z:\app\target\debug\build\napi-sys-9f2ceac139b31565\out\node-v12.19.0\include\node\node_api.h:225:33: error: unknown type name 'napi_threadsafe_function_call_js', err: true
  Z:\app\target\debug\build\napi-sys-9f2ceac139b31565\out\node-v12.19.0\include\node\node_api.h:226:33: error: unknown type name 'napi_threadsafe_function', err: true
  Z:\app\target\debug\build\napi-sys-9f2ceac139b31565\out\node-v12.19.0\include\node\node_api.h:229:38: error: unknown type name 'napi_threadsafe_function', err: true
  Z:\app\target\debug\build\napi-sys-9f2ceac139b31565\out\node-v12.19.0\include\node\node_api.h:233:31: error: unknown type name 'napi_threadsafe_function', err: true
  Z:\app\target\debug\build\napi-sys-9f2ceac139b31565\out\node-v12.19.0\include\node\node_api.h:235:31: error: unknown type name 'napi_threadsafe_function_call_mode', err: true
  Z:\app\target\debug\build\napi-sys-9f2ceac139b31565\out\node-v12.19.0\include\node\node_api.h:238:34: error: unknown type name 'napi_threadsafe_function', err: true
  fatal error: too many errors emitted, stopping now [-ferror-limit=], err: true
  thread 'main' panicked at 'Unable to generate napi bindings: ()', C:\Users\Ashcon\.cargo\git\checkouts\napi-rs-052f838a107f387b\f56852b\sys\build.rs:69:6
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
warning: build failed, waiting for other jobs to finish...
error: build failed
npm ERR! code ELIFECYCLE
npm ERR! errno 101
npm ERR! ovd@0.0.1 build-win: `cargo +stable-i686-pc-windows-msvc build && napi build`
npm ERR! Exit status 101
npm ERR!
npm ERR! Failed at the ovd@0.0.1 build-win script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\Ashcon\AppData\Roaming\npm-cache\_logs\2020-11-04T20_46_49_711Z-debug.log
```

**Comment by Brooooooklyn** [comment]
The `stable-i686-pc-windows-msvc` target passed on the CI, I think it was because of the host operation system on CI is `x64`. So I will create a `Windows x32` virtual machine to fix this issue, but it will cost times.

**Comment by rnd-ash** [comment]
That's an interesting note...

My setup is 64bit windows, with 32bit nodejs 14 installed. Had to install 32bit LLVM to compile napi since 64bit DLL was apparently invalid.

I looked at the action workflow and I'm confused as to why napi was able to compile with the 32bit target using a 64bit version of LLVM.

I'll setup a GitHub action for my codebase so we can compare build outputs. - the issue appears to be with napi-sys more than anything else according to my build logs.

**Comment by rnd-ash** [comment]
Can confirm. Something is wrong with your build process.
If i use the 64bit LLVM with my workflow it errors out:
https://github.com/rnd-ash/OpenVehicleDiag/actions/runs/347323659

**Comment by Brooooooklyn** [comment]
I will try again tonight

**Comment by adumbidiot** [comment]
@rnd-ash That error seems pretty self-explanatory. You are trying to run 32bit bindgen with a 64bit dll. You need to build as x64, but set the target to i686. 
See: https://github.com/napi-rs/napi-rs/blob/master/.github/workflows/windows-i686.yml

**Comment by adumbidiot** [comment]
Also, `bindgen` resolves `size_t` types to integers. napi uses this in its public abi. `Bindgen` doesn't use `usize` for this as it techinically isn't always equal to a `size_t`, though we can [configure `bindgen` to assume they are](https://docs.rs/bindgen/0.55.1/bindgen/struct.Builder.html#method.size_t_is_usize) since this is the case for most platforms.

**Comment by rnd-ash** [comment]
@adumbidiot the issue is I still need it to be built as 32bit....Installing 32bit LLVM on my windows PC and building that way gives me [these](https://github.com/napi-rs/napi-rs/issues/284#issuecomment-721967205) errors instead.

Reason for building 32bit...The API I'm using (J2534) is only 32bit, so all Windows DLLs that use it are 32bit only.

**Comment by adumbidiot** [comment]
You can set the output to be 32 bit while building everything else as 64 bit by using `--target`. Do you need the build scripts to be 32 bit too?

**Comment by rnd-ash** [comment]
Fixed that, now it seems we have a lot of linker issues:
https://github.com/rnd-ash/OpenVehicleDiag/actions/runs/349299626

**Comment by adumbidiot** [comment]
I managed to get it to [work](https://github.com/adumbidiot/OpenVehicleDiag/runs/1366415588) by 
1. Updating `napi-build` to be a git dep.
2. Hacking around `napi-rs` (the js module) since the version on npm is not updated enough and I couldn't get it to work as a git dep.

`napi-build` used to ignore a lot of important build options, but it doesn't do that on master anymore. the `napi` cli also got improvements for cross compilation.

**Comment by rnd-ash** [comment]
Interesting, I applied your changes @adumbidiot and Yes, it now compiles on github.

However when trying to install it on my own PC, NAPI-Sys fails:
```
warning: 2 warnings emitted

   Compiling napi-sys v0.4.7 (https://github.com/napi-rs/napi-rs#f56852bb)
error: failed to run custom build command for `napi-sys v0.4.7 (https://github.com/napi-rs/napi-rs#f56852bb)`

Caused by:
  process didn't exit successfully: `Z:\app\target\debug\build\napi-sys-1339bd91616aeb4c\build-script-build` (exit code: 101)
  --- stdout
  cargo:rerun-if-env-changed=NODE_INCLUDE_PATH
  cargo:rerun-if-changed=src\bindings.h
  cargo:rerun-if-changed=src\lib.rs
  cargo:rerun-if-changed=src\sys.rs

  --- stderr
  thread 'main' panicked at 'Unable to find libclang: "couldn\'t find any valid shared libraries matching: [\'clang.dll\', \'libclang.dll\'], set the `LIBCLANG_PATH` environment variable to a path where one of these files can be found (invalid: [(C:\\Program Files\\LLVM\\bin\\libclang.dll: invalid DLL (64-bit))])"', C:\Users\Ashcon\.cargo\registry\src\github.com-1285ae84e5963aae\bindgen-0.55.1\src/lib.rs:1896:31
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
PS Z:\app>
```

I was following the github actions script line by line...so what gives??

**Comment by Brooooooklyn** [comment]
Do you forget to set LIBCLANG_PATH ?

**Comment by adumbidiot** [comment]
> `invalid: [(C:\\Program Files\\LLVM\\bin\\libclang.dll: invalid DLL (64-bit))]`
@rnd-ash You sure you set the `--target` flag correctly when building?

**Comment by rnd-ash** [comment]
@Brooooooklyn, Yes, LIBCLANG path is set to C:\Program Files\LLVM\bin
And I am using the build-win build script from package.json as per @adumbidiot's version.

What I don't get is that it seems Github is also installing a 64bit version of LLVM via choco and that seems to build correctly...

**Comment by adumbidiot** [comment]
I don't think its a powershell issue as the `LIB_CLANG` var appears to have been recognized; the build script is simply built as 32bit so it cant use the located dll. What command did you use to build the rustlib?

**Comment by rnd-ash** [comment]
I used `cargo build --target i686-pc-windows-msvc` to attempt to build the library, same command thats used in package.json which the Github action seems to complete without a problem

**Comment by adumbidiot** [comment]
Try running `rustup default stable` and try building again.

**Comment by rnd-ash** [comment]
That seemed to fix the build issue! @adumbidiot 

However, now when trying to load the library, I get this:
```
> let lib=require('./index.node');

Uncaught Error: Access is denied.
\\?\Z:\app\index.node
    at Object.Module._extensions..node (internal/modules/cjs/loader.js:1122:18)
    at Module.load (internal/modules/cjs/loader.js:928:32)
    at Function.Module._load (internal/modules/cjs/loader.js:769:14)
    at Module.require (internal/modules/cjs/loader.js:952:19)
    at require (internal/modules/cjs/helpers.js:88:18)
```

This only occurs on Windows...

**Comment by rnd-ash** [comment]
Re-installing node as x86 now gives me this:
```
Uncaught Error: \\?\Z:\app\index.node is not a valid Win32 application.
\\?\Z:\app\index.node
    at process.func [as dlopen] (electron/js2c/asar.js:140)
    at Object.Module._extensions..node (internal/modules/cjs/loader.js:1196)
    at Object.func [as .node] (electron/js2c/asar.js:140)
    at Module.load (internal/modules/cjs/loader.js:981)
    at Module._load (internal/modules/cjs/loader.js:881)
    at Function.Module._load (electron/js2c/asar.js:769)
    at Module.require (internal/modules/cjs/loader.js:1023)
    at require (internal/modules/cjs/helpers.js:77)
    at selector.js:1

```

**Comment by adumbidiot** [comment]
@rnd-ash You probably have 64 bit electron installed and it can't load 32 bit modules. See https://www.electronjs.org/docs/tutorial/installation#customization. You can check your currently installed electron's process through the devtools console with something like `process.arch`.

**Comment by rnd-ash** [comment]
Awesome! It works now!,

I assume this bug should still be left open due to having to hack round napi-build with a custom node command right?

**Comment by Brooooooklyn** [comment]
> I assume this bug should still be left open due to having to hack round napi-build with a custom node command right?

Yes, I will fix it soon.

**Comment by adumbidiot** [comment]
I thought that was fixed on master? I just couldn't get projects to depend on a git version of the cli

**Comment by Brooooooklyn** [comment]
Just published `napi-rs@v0.3.10` from master branch @rnd-ash
